File: typescript/src/middleware/inference.ts
import type { Message, ToolDefinition } from "../llm/request.js";
import type { Response, ToolCall as LlmToolCall } from "../llm/response.js";
import type { Backend } from "../llm/backend.js";
import type { ContextWindowManager } from "../llmcontext/context_window.js";
import { Role } from "../llm/request.js";
import { MessageType } from "./types.js";
import type { ResponseValidator, ToolCall, ValidationResult } from "./guardrails/response_validator.js";
import { ErrorTracker, ErrorCategory } from "./guardrails/error_tracker.js";
import { StepEnforcer } from "./guardrails/step_enforcer.js";
import { stepNudge } from "./guardrails/nudge.js";

export class RetriesExhaustedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "RetriesExhaustedError";
  }
}

export interface InferenceResult {
  response: Response;
  newMessages: Message[];
  toolCallCounter: number;
  attempts: number;
}

export interface InferenceConfig {
  client: Backend;
  contextManager?: ContextWindowManager;
  validator?: ResponseValidator;
  errorTracker?: ErrorTracker;
  stepEnforcer?: StepEnforcer;
  toolSpecs: ToolDefinition[];
  maxAttempts: number;
  stepIndex?: number;
}

function getToolNames(specs: ToolDefinition[]): string[] {
  return specs.map((spec) => spec.name);
}

export async function runInference(
  messages: Message[],
  cfg: InferenceConfig,
  sessionId: string = ""
): Promise<InferenceResult | null> {
  let maxAttempts = cfg.maxAttempts;
  if (maxAttempts <= 0) {
    maxAttempts = 3;
  }

  const currentMessages: Message[] = [...messages];

  let lastResponse: Response | null = null;
  let toolCallCounter = 0;
  let attempts = 0;

  while (attempts < maxAttempts) {
    attempts++;

    if (cfg.contextManager) {
      if (cfg.contextManager.shouldCompact(currentMessages)) {
        const compacted = cfg.contextManager.compact(currentMessages);
        currentMessages.length = 0;
        currentMessages.push(...compacted);
      }

      const warning = cfg.contextManager.checkThresholds(currentMessages);
      if (warning) {
        const warningMsg: Message = {
          role: Role.USER,
          content: warning,
          meta: { type: MessageType.CONTEXT_WARNING },
        };
        currentMessages.push(warningMsg);
      }
    }

    let resp: Response;
    try {
      resp = cfg.client.complete(currentMessages);
    } catch (e) {
      if (cfg.errorTracker && e instanceof Error) {
        cfg.errorTracker.recordError(sessionId, "", {}, e, ErrorCategory.UNKNOWN);
      }
      throw e;
    }

    if (cfg.contextManager && resp.usage_tokens.total_tokens > 0) {
      cfg.contextManager.updateTokenCount(resp.usage_tokens.total_tokens);
    }

    let validationResult: ValidationResult | null = null;

    if (resp.tool_calls && resp.tool_calls.length > 0) {
      const guardrailsToolCalls: ToolCall[] = resp.tool_calls.map((tc) => ({
        tool: tc.name,
        args: tc.arguments,
      }));
      if (cfg.validator) {
        validationResult = cfg.validator.validateToolCalls(guardrailsToolCalls);
      } else {
        validationResult = {
          toolCalls: guardrailsToolCalls,
          nudge: null,
          needsRetry: false,
        };
      }
    } else if (resp.content) {
      if (cfg.validator) {
        validationResult = cfg.validator.validateTextResponse(resp.content);
      } else {
        validationResult = { toolCalls: [], nudge: null, needsRetry: false };
      }

      if (!validationResult.needsRetry && validationResult.toolCalls.length > 0) {
        resp.tool_calls = validationResult.toolCalls.map(
          (tc) =>
            ({
              id: "",
              name: tc.tool,
              arguments: tc.args,
            } as LlmToolCall)
        );
      }
    } else {
      if (cfg.validator) {
        validationResult = cfg.validator.validateTextResponse("");
      } else {
        validationResult = { toolCalls: [], nudge: null, needsRetry: true };
      }
    }

    lastResponse = resp;

    if (validationResult && !validationResult.needsRetry) {
      if (cfg.errorTracker) {
        cfg.errorTracker.resetSession(sessionId);
      }

      if (cfg.stepEnforcer && resp.tool_calls && resp.tool_calls.length > 0) {
        for (const tc of resp.tool_calls) {
          const [allowed, missing] = cfg.stepEnforcer.canExecute(sessionId, tc.name);
          if (!allowed) {
            const nudge = stepNudge(tc.name, missing, 1);
            const nudgeMsg: Message = {
              role: Role.USER,
              content: nudge.content,
              meta: { type: MessageType.STEP_NUDGE },
            };
            currentMessages.push(nudgeMsg);
            continue;
          }
        }
      }

      toolCallCounter += resp.tool_calls ? resp.tool_calls.length : 0;
      return {
        response: lastResponse,
        newMessages: currentMessages,
        toolCallCounter,
        attempts,
      };
    }

    if (cfg.errorTracker) {
      cfg.errorTracker.recordError(
        sessionId,
        "",
        {},
        new Error("validation failed"),
        ErrorCategory.UNKNOWN
      );
    }

    if (attempts >= maxAttempts) {
      throw new RetriesExhaustedError(`retries exhausted after ${attempts} attempts`);
    }

    if (validationResult && validationResult.nudge) {
      const nudgeMsg: Message = {
        role: Role.USER,
        content: validationResult.nudge.content,
        meta: { type: MessageType.RETRY_NUDGE },
      };
      currentMessages.push(nudgeMsg);
    }

    const failedMsg: Message = {
      role: Role.ASSISTANT,
      content: resp.content,
      meta: { type: MessageType.TEXT_RESPONSE },
    };
    currentMessages.push(failedMsg);
  }

  throw new RetriesExhaustedError(`retries exhausted after ${attempts} attempts`);
}

File: python/src/pedro_agentware/middleware/inference.py
"""Inference loop wiring guardrails together."""

import asyncio
from dataclasses import dataclass, field

from pedro_agentware.llm import Message, Role
from pedro_agentware.llm.backend import Backend
from pedro_agentware.llm.request import ToolDefinition
from pedro_agentware.llm.response import Response
from pedro_agentware.llm.response import ToolCall as LLMToolCall
from pedro_agentware.llmcontext.context_window import ContextWindowManager
from pedro_agentware.middleware.guardrails.error_tracker import ErrorCategory, ErrorTracker
from pedro_agentware.middleware.guardrails.response_validator import (
    ResponseValidator,
    ValidationResult,
)
from pedro_agentware.middleware.guardrails.response_validator import (
    ToolCall as GuardrailsToolCall,
)
from pedro_agentware.middleware.guardrails.step_enforcer import StepEnforcer
from pedro_agentware.middleware.types import MessageMeta, MessageType


class RetriesExhaustedError(Exception):
    """Raised when inference attempts are exhausted."""

    pass


@dataclass
class InferenceResult:
    """Result of an inference call."""

    response: Response
    new_messages: list[Message]
    tool_call_counter: int
    attempts: int


@dataclass
class InferenceConfig:
    """Configuration for the inference loop."""

    client: Backend
    context_manager: ContextWindowManager | None = None
    validator: ResponseValidator | None = None
    error_tracker: ErrorTracker | None = None
    step_enforcer: StepEnforcer | None = None
    tool_specs: list[ToolDefinition] = field(default_factory=list)
    max_attempts: int = 10
    step_index: int = 0


def _get_tool_names(specs: list[ToolDefinition]) -> list[str]:
    """Extract tool names from tool specifications."""
    return [spec.name for spec in specs]


async def run_inference(
    messages: list[Message],
    cfg: InferenceConfig,
    session_id: str = "",
) -> InferenceResult | None:
    """Run inference loop with guardrails.

    Args:
        messages: Current conversation messages.
        cfg: Inference configuration.
        session_id: Session identifier for step enforcer and error tracker.

    Returns:
        InferenceResult on success, None if all attempts exhausted without valid response.

    Raises:
        RetriesExhaustedError: When max attempts are reached without a valid response.
    """
    if cfg.max_attempts <= 0:
        cfg.max_attempts = 3

    current_messages = list(messages)

    last_response: Response | None = None
    tool_call_counter = 0
    attempts = 0

    while attempts < cfg.max_attempts:
        attempts += 1

        if cfg.context_manager is not None:
            if cfg.context_manager.should_compact(current_messages):
                compacted = cfg.context_manager.compact(current_messages)
                current_messages = compacted

            warning = cfg.context_manager.check_thresholds(current_messages)
            if warning:
                warning_msg = Message(
                    role=Role.USER,
                    content=warning,
                    meta=MessageMeta(type=MessageType.CONTEXT_WARNING),
                )
                current_messages.append(warning_msg)

        tool_specs_serialized = []
        for spec in cfg.tool_specs:
            tool_specs_serialized.append({
                "type": "function",
                "function": {
                    "name": spec.name,
                    "description": spec.description,
                    "parameters": spec.input_schema,
                },
            })

        try:
            resp = await asyncio.get_event_loop().run_in_executor(
                None,
                lambda: cfg.client.complete(current_messages),
            )
        except Exception as e:
            if cfg.error_tracker is not None:
                cfg.error_tracker.record_error(
                    session_id,
                    "",
                    {},
                    e,
                    ErrorCategory.UNKNOWN,
                )
            raise

        if cfg.context_manager is not None and resp.usage_tokens.total_tokens > 0:
            cfg.context_manager.update_token_count(resp.usage_tokens.total_tokens)

        validation_result: ValidationResult | None = None

        if resp.tool_calls:
            guardrails_tool_calls = [
                GuardrailsToolCall(tool=tc.name, args=tc.arguments)
                for tc in resp.tool_calls
            ]
            if cfg.validator:
                validation_result = cfg.validator.validate_tool_calls(guardrails_tool_calls)
            else:
                validation_result = ValidationResult(
                    tool_calls=guardrails_tool_calls, nudge=None, needs_retry=False
                )
        elif resp.content:
            if cfg.validator:
                validation_result = cfg.validator.validate_text_response(resp.content)
            else:
                validation_result = ValidationResult(tool_calls=[], nudge=None, needs_retry=False)

            if not validation_result.needs_retry and validation_result.tool_calls:
                resp.tool_calls = [
                    LLMToolCall(id="", name=tc.tool, arguments=tc.args)
                    for tc in validation_result.tool_calls
                ]
        else:
            if cfg.validator:
                validation_result = cfg.validator.validate_text_response("")
            else:
                validation_result = ValidationResult(
                    tool_calls=[],
                    nudge=None,
                    needs_retry=True,
                )

        last_response = resp

        if validation_result and not validation_result.needs_retry:
            if cfg.error_tracker is not None:
                cfg.error_tracker.reset_session(session_id)

            if cfg.step_enforcer is not None and resp.tool_calls:
                for tc in resp.tool_calls:
                    allowed, missing = cfg.step_enforcer.can_execute(session_id, tc.name)
                    if not allowed:
                        from pedro_agentware.middleware.guardrails.nudge import step_nudge

                        nudge = step_nudge(tc.name, missing, 1)
                        nudge_msg = Message(
                            role=Role.USER,
                            content=nudge.content,
                            meta=MessageMeta(type=MessageType.STEP_NUDGE),
                        )
                        current_messages.append(nudge_msg)
                        continue

            tool_call_counter += len(resp.tool_calls) if resp.tool_calls else 0
            return InferenceResult(
                response=last_response,
                new_messages=current_messages,
                tool_call_counter=tool_call_counter,
                attempts=attempts,
            )

        if cfg.error_tracker is not None:
            cfg.error_tracker.record_error(
                session_id,
                "",
                {},
                Exception("validation failed"),
                ErrorCategory.UNKNOWN,
            )

        if attempts >= cfg.max_attempts:
            raise RetriesExhaustedError(f"retries exhausted after {attempts} attempts")

        if validation_result and validation_result.nudge:
            nudge_msg = Message(
                role=Role.USER,
                content=validation_result.nudge.content,
                meta=MessageMeta(type=MessageType.RETRY_NUDGE),
            )
            current_messages.append(nudge_msg)

        failed_msg = Message(
            role=Role.ASSISTANT,
            content=resp.content,
            meta=MessageMeta(type=MessageType.TEXT_RESPONSE),
        )
        current_messages.append(failed_msg)

    raise RetriesExhaustedError(f"retries exhausted after {attempts} attempts")
